backend/remote-state/consul/client.go

```diff
@@ -2,15 +2,29 @@ package consul
 
 import (
 	"crypto/md5"
+	"encoding/json"
+	"errors"
+	"time"
 
 	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/errwrap"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform/state"
 	"github.com/hashicorp/terraform/state/remote"
 )
 
+const (
+	lockSuffix     = "/.lock"
+	lockInfoSuffix = "/.lockinfo"
+)
+
 // RemoteClient is a remote client that stores data in Consul.
 type RemoteClient struct {
 	Client *consulapi.Client
 	Path   string
+
+	consulLock *consulapi.Lock
+	lockCh     <-chan struct{}
```

Comment:
Concurrent calls of Lock would definitely be a bug. Lock implementations aren't guaranteed to be reentrant, so we can't ever count on multiple Lock calls working.

I think it's better to ensure that we internally maintain unambiguous locations for taking and releasing the lock, rather than sprinkling Unlock around to hope it gets called.